prov/netdir: Added tests for Network Direct provider

gladkovdmitry17 opened the pull request:
1) NetworkDirect provider functionality are covered in fabtests
    - Basic MSG functionality (send/recv)
    - Basic RMA functionality (write/read)
    - Large message transmisison via send
    - Cancellation of recv events
    - send/recv with multiple iovecs
    - read/write with multiple iovecs
2)  The functionality of RXM based on NetworkDirect are covered in fabtests
    - Basic functionality (send/recv)
    - Tagged operations (tsend/trecv)
    - Tagged operation (tinject)
    - Check possibility to use FI_SELECTIVE_COMPLETION for CQ and 
       fi_tx/rx_attr::op_flags = FI_COMPLETION

Signed-off-by: Dmitry Gladkov <dmitry.gladkov@intel.com>

shefty commented:
This is a long list of changes for a single commit.  Please limit each patch to a single change as much as possible.

shefty commented on common/shared.c:
```diff
@@ -833,11 +833,14 @@ void ft_free_res(void)
+	if (rx_ctx_arr) {
+		free(rx_ctx_arr);
+		rx_ctx_arr = NULL;
+	}
```
It is safe to pass a null pointer to free().  Does Windows not allow this?

gladkovdmitry17 replied:
Yes, it's safe too. It was done just to align with "free" above. I'll revert this change

shefty commented on getopt/getopt.c:
```diff
@@ -0,0 +1,713 @@
+typedef int bool;
+
+#define true 1
+#define false 0
```
Let's not define bool, true, or false.

gladkovdmitry17 replied:
Ok

shefty commented on getopt/getopt.c:
```diff
@@ -0,0 +1,713 @@
+
+
+
+
```
The formatting of this file does not match the rest of the code base.  There's also no license at the top of the code.  We need to know the origin of this code and what license it had.

If this code is only needed for Windows, we may want to move it to an OS specific directory.

gladkovdmitry17 replied:
I'm going to replace getopt.[ch] files to **osd** directory. I guess it's right. Please, correct me if I'm wrong.
The formatting will be fixed and aligned with the rest of the code base.
Sorry, I forgot about. I'll add license header for both files.

gladkovdmitry17 replied:
Sorry, not a **osd** directory, it's just a filter in VS. getopt.[ch] will be moved to include/windows/getopt folder under GIT vc

shefty commented on getopt/getopt.h:
```diff
@@ -0,0 +1,57 @@
+#endif __cplusplus
+
+#endif //_GETOPT_H_
+
```
Same comments as getopt.c above.

gladkovdmitry17 replied:
I'm applying the same as for getopt.c

shefty commented on include/windows/osd.h:
```diff
@@ -0,0 +1,168 @@
+    QueryPerformanceCounter(&count);
+
+    tp->tv_sec = (time_t)((double)count.QuadPart / freq.QuadPart);
+    tp->tv_nsec = (long)((uint64_t)((double)count.QuadPart * 1000000 / freq.QuadPart) % 1000000);
```
Please watch line length, and try to keep to around 80 characters based on readability.

gladkovdmitry17 replied:
Ok, I'll check other places too

shefty commented on include/windows/osd.h:
```diff
@@ -0,0 +1,168 @@
+{
+    char* res = strdup(str);
+    if(strlen(res) > n)
+        res[n] = 0;
```
Use '\0' instead

gladkovdmitry17 replied:
Ok

shefty commented on include/windows/osd.h:
```diff
@@ -0,0 +1,168 @@
+    closesocket(socks[1]);
+    WSASetLastError(err);
+    return SOCKET_ERROR;
+}
```
This function is a little much for a static inline.  Please move into a .c file

gladkovdmitry17 replied:
I'm going to create osd.c file that will contain the function

shefty commented on simple/msg_netdir.c:
```diff
@@ -0,0 +1,870 @@
+
+	return 0;
+}
+#endif /* !_WIN32 */
```
fabtests should not have provider specific tests.  Any test should work over any provider that supports the necessary capabilities.  Please rename this test and submit separately if it needs to be included.  The name needs to reflect what the test is attempting to verify.  The test should do something not covered by other tests, which I'm not sure this does.

If you need to update the other tests to work over the ND provider (e.g. hook in WSAStartup somehow), please do that before adding a new test.

shefty commented on simple/msg_sockets.c:
```diff
@@ -306,6 +306,13 @@ static int setup_handle(void)
+	ret = WSAStartup(MAKEWORD(2, 2), &data);
+	if (ret)
+		return HRESULT_FROM_WIN32(ret);
+#endif
```
Let's add some sort of ft_init() or ft_startup() call to do this, and have the tests call that instead of adding WSAStartup to each test like this.

gladkovdmitry17 replied:
Ok, I'm going to add ft_startup with appropriate stuff inside.

shefty commented on simple/poll.c:
```diff
@@ -216,7 +216,7 @@ int main(int argc, char **argv)
 		return EXIT_FAILURE;
 
-	while ((op = getopt(argc, argv, "h" ADDR_OPTS CS_OPTS INFO_OPTS)) != -1) {
+	while ((op = getopt(argc, argv, "h" CS_OPTS INFO_OPTS)) != -1) {
```
We need to explain this change.

gladkovdmitry17 replied:
ADDR_OPTS is alreadyh included in CS_OPTS. This chagne is to just to avoid double ADDR_OPTS in getopt's options string

shefty commented on simple/rxm_netdir.c:
```diff
@@ -0,0 +1,484 @@
+
+	return 0;
+}
+#endif /* !_WIN32 */
```
Same comment applies as to the other ND specific test.

gladkovdmitry17 commented:
@shefty Thanks for the review changes of the PR. About you comment to "limit each patch into a single change" - Do you mean create several commits (for example, 
1. Simple MSG NetDir tests
2. Sinple RMA NertDir tests
2. NetDir + RXM
3. Some fixes) ?

shefty commented:
> @shefty <https://github.com/shefty>  Thanks for the review changes of
> the PR. About you comment to "limit each patch into a single change" -
> Do you mean create several commits (for example,
> 
> 1.	Simple MSG NetDir tests
> 2.	Sinple RMA NertDir tests
> 3.	NetDir + RXM
> 4.	Some fixes) ?

Yes - separating the changes makes it easier to review and merge.  thanks

a-ilango commented:
Can the tests be made provider independent? Some of the tests like send_recv_greeting is already covered under simple/msg.c. Also some functions that establishes the connection like client_connect, server_connect  etc have already been implemented in common/shared.c. They can be reused.

gladkovdmitry17 commented:
@a-ilango yes, I've done that in cases where it's possible (please, check it for msg_netdir.c source file). The following modification are done:
- server_listen -> ft_start_server
- server_connect -> ft_server_connect
- client_connect -> ft_client_connect

Other functions that are responsible for transmit/receive msg can not be replaced since it's just implemented specific test cases. Btw, send_recv_greeting is already used in msg_netdir.

As for rdm_netdir.c source file, setup_handle -> ft_init_fabric

gladkovdmitry17 commented:
This content was splitted into 4 commits:
- Added some stuff to support fabtests in Win OS
- Covered NetDir
- Changed ft_inject to support address as a parameter, minor fix in poll.c, added WSAStartup in msg_sockets.c (added ft_startup func)
- Covered RXM over NetDir

a-ilango commented on common/shared.c:
```diff
@@ -213,123 +213,25 @@ static void ft_cntr_set_wait_attr(void)
-	free(combo);
+	return mr_access;
 }
 
```
@gladkovdmitry17 This patch removes a lot of code above and it looks like a side effect of not pulling in latest changes from fabtests master.

gladkovdmitry17 replied:
@a-ilango in 4th commit, all missing stuff was added.
I'm going to split each added functionality into separate commit tomorrow.

a-ilango replied:
Thanks. A commit shouldn't break the build. So please ensure each commit doesn't remove existing features.

gladkovdmitry17 replied:
Thank you too! Sean advised me how PR should be prepared. All changes outside ND provider will be put into separate PR and can be simple reviewed

gladkovdmitry17 commented:
The PR has been updated. Only NetworkDirect and RXM+NetworkDirect tests  are stayed.

shefty commented:
Why do we need to add 1200+ lines of fabtests that only work with one provider?  Why can't these tests either be added to an existing test or added as generic API tests?

gladkovdmitry17 commented:
@shefty Sure, these tests can be added as a part of additional tests to already existing test cases. Please, let me a couple of days, I'll add these tests to already existing test code.
Thank you!

shefty commented:
Closing this, so that more generic tests can be used/expanded for the ND provider. There's no rush adding these changes to an existing tests, so just open a new PR when (or if) they are ready.  Thanks!

gladkovdmitry17 commented:
@shefty Sure. Thanks!
